### ISSUE TITLE
Imporve redundant_slicing lint by taking the expected type into account

### ISSUE THREAD

**Issue by YohDeadfall** [rust-lang/rust-clippy#7257]
Lint name: `redundant_slicing`

I tried this code:

```rust
fn main() {
    let v = vec![1, 2, 3];
    test_vec(&v[..]);

    let v = [1, 2, 3];
    test_vec(&v[..]);

    let s = "Hello, World!";
    test_str(&s[..]);

    let s = "Hello, World!".to_string();
    test_str(&s[..]);
}

fn test_vec(_: &[i32]) {
}

fn test_str(_: &str) {
}
```

I expected to see a warning for each call, but instead only 9th is reported since types before and after indexing are same. It would be much better if clippy is able to see what type is expected by a higher expression/parameter.

If it's possible, I would like to do it myself.

### Meta

- `cargo clippy -V`: clippy 0.1.53 (bacf770f 2021-05-05)
- `rustc -Vv`:
  ```
  rustc 1.54.0-nightly (bacf770f2 2021-05-05)
  binary: rustc
  commit-hash: bacf770f2983a52f31e3537db5f0fe1ef2eaa874
  commit-date: 2021-05-05
  host: x86_64-pc-windows-msvc
  release: 1.54.0-nightly
  LLVM version: 12.0.0
  ```

**Comment by Jarcho** [comment]
You're essentially checking if the result of dereferencing the value would have the same type as slicing. You'll also need to check to make sure auto-deref will happen when the change is made (this is the hard part).

**Comment by YohDeadfall** [comment]
I found the algorithm used by auto deref, but what I can't understand is how to get traits implemented by a struct from a `LateContext`.

**Comment by giraffate** [comment]
I don't know the details of the implementation, but FYI this might be useful: https://github.com/rust-lang/rust-clippy/blob/master/doc/common_tools_writing_lints.md#checking-if-a-type-implements-a-specific-trait
